src/graphforge/planner/operators.py

```
@@ -76,6 +104,59 @@ def validate_direction(cls, v: str) -> str:
+    edge_types: list[str] = Field(..., description="Edge types to match")
+    direction: str = Field(..., description="Traversal direction")
+    min_hops: int = Field(default=1, description="Minimum hops")
+    max_hops: int | None = Field(default=None, description="Maximum hops (None=unbounded)")
+
+    @field_validator("direction")
+    @classmethod
+    def validate_direction(cls, v: str) -> str:
+        """Validate direction is valid."""
+        valid_dirs = {"OUT", "IN", "UNDIRECTED"}
+        if v not in valid_dirs:
+            raise ValueError(f"Direction must be one of {valid_dirs}, got {v}")
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)